kernel: arch: soc: introduce 'k_entry' symbol

romkell opened the pull request:
# Issue

Zephyr does not provide a common compiler symbol usable as entry point for a bootloader to jump to.
The best comparable is the string behind `CONFIG_KERNEL_ENTRY` which is also used in the linker script `ENTRY(CONFIG_KERNEL_ENTRY)`

# User Story / Use Case

I want to have a common compiler/linker symbol with the address of the kernel entry point (equivalent to __start, __nuclei_start, entry, _MainEntry etc.) which I can place into a (const) variable `entry_point` in a firmware information block accessible by a bootloader, for the bootloader to simply jump to that address location with a function pointer with signature `void (*entry_point)(void)`

# Solution
- Define a common `k_entry` symbol in each architecture reset.S, start.S, entry.S etc. assembly file, making `k_entry` an equivalent to the different __start, entry, __nuclei_start etc. symbols by
  - Setting the CONFIG_KERNEL_ENTRY to the correct symbol e.g. __nuclei_start, rom_start, etc.
  - Making the content of CONFIG_KERNEL_ENTRY an "alias" symbol by using the PROVIDE() directive in the linker script.
- Declare (void)k_entry(void) as public function in a kernel header file

# Limitation

Not all cores are covered by this PR (mainly ESP32 with appcpu and procpu) and will have to be solved individually.

nashif commented:
we are working on removing usage of z_ as a prefix, please come up with a different name. Also, are you working on fixing the CI issues?

romkell commented:
> we are working on removing usage of z_ as a prefix, please come up with a different name. Also, are you working on fixing the CI issues?

I found that when search for naming conventions:

- https://github.com/zephyrproject-rtos/zephyr/wiki/Naming-Conventions
- https://github.com/zephyrproject-rtos/zephyr/issues/29569

So if that is still valid, I suggest either
- `k_entry` or
- `k_kernel_entry` a bit of a duplication for `k_` for kernel and `kernel` for kernel but closer to `KCONFIG_KERNEL_ENTRY`

A I see it, it is a kernel feature and it is exposed in a include, hence `k_` should be the one to choose.
Since `k_` marks it as kernel API I would rather go with the short form.

Fixing: As soon as I get around to fix it, yes. I converted into a draft MR.

dcpleung commented on soc/amd/acp_6_0/Kconfig.defconfig:
```diff
@@ -20,6 +20,6 @@ config MULTI_LEVEL_INTERRUPTS
 	default n
 
-config KERNEL_ENTRY
+config CUSTOM_SOC_KERNEL_ENTRY
 	default "__start"
```
This is defined as boolean so this needs to be changed.

dcpleung reviewed (commented):
Are you building the bootloader and the app together? From the description it sounded like the bootloader was built separately, and that's why I suggested using ELF file to grab the entry point. You can build the app first, grab the entry point address and feeds it into the bootloader building process. If they are built together, you can use the entry point string directly, and let the linker fill the address for you.

One issue with this type of intrusive changes across the tree is that, there is no way to test everything as CI does not cover anything, especially that it does not test on any real hardware. And this particularly affects the boot process. If this breaks on a board, it is going to be very confusing to the developers as it simply no longer boots without any error messages. So one must be very careful on this type of changes.

I recommend going to the arch meeting or the dev meeting to present this PR and to gain feedback, and also to give a heads-up to others before this is merged.

romkell committed:
sys: tracing: quick fix to get compiled

Make k_event tracing macros void to allow compilation in order to make
pipeline succeed.

Commit to be dropped, when when
fixed.

Signed-off-by: Roman Kellner <rkellner@baumer.com>

andyross reviewed (commented):
No real complaints, but this seems like a lot of churn for very little value.  It's not really possible in the general case to have a "Zephyr" entry point.  Every system boots differently, and some systems enter in one of N ways based on configuration (think about TrustZone configurations, or zephyr-under-bootloader vs. zephyr-as-bootloader, also the Xtensa audio DSPs almost all manage their own entry code independent of the arch layer because of the complexities of host-managed startup).

I mean, there's nothing wrong with providing a "k_entry" symbol, exactly.  But really we'll never be able to do anything portable with it.    No vote, but my gut says that this wants to be submitted at a per-SOC level.

mathieuchopstm commented:
\+1 on what @andyross said. If that's something you ***need***, nothing prevents you from adding a custom linker snippet like `PROVIDE(__my_kernel_entrypoint = CONFIG_KERNEL_ENTRY)` via the various CMake linker script functions.

dcpleung commented:
If that's the case with your bootloader setup, I think it would actually be better to have a fixed address for the bootloader to jump to. The app linker script then put the entry point at that same address. With this, there would be less ambiguity.

romkell commented:
> No real complaints, but this seems like a lot of churn for very little value. 

"... a lot of churn ..." is probably the nature of changing in the very core, since affecting all arch's.

"little value" is in the eye of the beholder.
Alternative would be:
- Make CONFIG_KERNEL_ENTRY a symbol - does not even work persistently, if the SOC integrater did not set CONFIG_KERNEL_ENTRY correctly e.g. `__appcpu_start`, `rom_start`, should no longer happen if `k_entry` is the one and only ld ENTRY() symbol.
- Use readelf & friends an patch the elf post compilation

> It's not really possible in the general case to have a "Zephyr" entry point. Every system boots differently, and some systems enter in one of N ways based on configuration (think about TrustZone configurations, or zephyr-under-bootloader vs. zephyr-as-bootloader, also the Xtensa audio DSPs almost all manage their own entry code independent of the arch layer because of the complexities of host-managed startup).

I am not an expert on all those the named platforms and configurations, but as long as there always is that one very first executed instruction in all system, no matter whether you get there via system reset and reset vector or jump there from a bootloader, to my understand that is `k_entry`.
As implemented by now, there would be 3 options for an SOC (`reset.S` here is just a placeholder for any arch common start-up code file):
- use the standard arch `reset.S` file and `__start`, if that is sufficient - the arch `reset.S` defines `k_entry` in that case
- prepend some SOC specific start-up code before jumping to the standard arch  `reset.S` file and `__start` and continue there - the SOC specific start-up file then defines `k_entry` (as was done with CONFIG_KERNEL_ENTRY which became `"__nuclei_start"`or `"entry"` instead of the default `"__start"`, (if not forgotten to be set in CONFIG_KERNEL_ENTRY ).
- implement your complete own start-up code, then define `k_entry` there, either do not compile `reset.S` or suppress the `k_entry`definition in the common/arch `reset.S`.

> 
> I mean, there's nothing wrong with providing a "k_entry" symbol, exactly. But really we'll never be able to do anything portable with it. 

No clear to me what you understand as `portable`. if `k_entry` is the address of the very first executed instruction and I can expose it to a bootloader (as described above) the contract would be "jump to or start from that address and all the rest is taken care of by the subsequent code". I would consider that portable. What did I not consider in lack of knowledge of e.g TrustZone or others?

>No vote, but my gut says that this wants to be submitted at a per-SOC level.

- Is that meant to split into various commits: infrastructure first, then a number of commit per arch and soc?
- Or split into multiple PRs (Hence a slow transition?)
  - add new infrastructure
  - then arch 1
  - then arch 2,
  - then arch N,
  - then remove old infrastructure?
  
The later I consider a bit dangerous, to get stuck in the middle of such the transition for some arch's no get review / accepted. I feel it should be one PR. Organizing commit makes sense.

romkell commented:
> +1 on what @andyross said. If that's something you _**need**_, nothing prevents you from adding a custom linker snippet like `PROVIDE(__my_kernel_entrypoint = CONFIG_KERNEL_ENTRY)` via the various CMake linker script functions.

Thanks. Good hint. Did not know `PROVIDE(__my_kernel_entrypoint = CONFIG_KERNEL_ENTRY)`.

Downside, CONFIG_KERNEL_ENTRY is not consistently used. Some SOCs have their entry point named `rom_start`, or `__appcpu_start` etc. have a linker script with `ENTRY(rom_start)` or `ENTRY(__appcpu_start)` but not `ENTRY(CONFIG_KERNEL_ENTRY)`. CONFIG_KERNEL_ENTRY is just left on its default value.
Yes, I then can do `PROVIDE(__my_kernel_entrypoint = __appcpu_start)` but would have to adjust for each arch.

romkell commented:
> If that's the case with your bootloader setup, I think it would actually be better to have a fixed address for the bootloader to jump to. The app linker script then put the entry point at that same address. With this, there would be less ambiguity.

We could put the entry point always to the same offset within the executable partition(s). But I still need to know the entry_points name which is not consistently used in CONFIG_KERNEL_ENTRY.  See comment below.

romkell commented:
> I might be completely off but throwing it out there just in case: maybe binary descriptors could be useful here? https://docs.zephyrproject.org/latest/services/binary_descriptors/index.html

Had a quick look at the docu.
Pretty similar what we plan to do for the firmware info and partition table. 32 bit magic at the front, 32 bit bit inverse magic at the end.

The single bindesc's will be too costly concerning flash due to the overhead.

What might work is `BINDESC_BYTES_DEFINE`.

Advantage would be having tools ready such as `west bindesc`.

Parsing for such sections is ok for tools, but, to us, not for a bootloader (boot time). We need a pointer to it.
The whole bindesc module must to be very lean for us. We had to replace the Zephyr `retention` module by an own implementation due to code size.

Possibly need to have a deeper look. Thanks for the hint.

romkell committed:
kernel: introduce 'k_entry' symbol

Introduce 'k_entry' as common entry point symbol, which can be
stored as function pointer address in a variable made accessible to a
bootloader. The bootloader then can use that address to start the
firmware image.
The k_entry "alias" symbol is created by the linker script PROVIDE()
directive using CONFIG_KERNEL_ENTRY as original symbol.

Signed-off-by: Roman Kellner <rkellner@baumer.com>

romkell commented on subsys/tracing/test/tracing_test.h:
@nashif @peter-mitsis : approaching you since I found you adding the other `sys_port_trace_k_event_<xyz>`.

I got compile error one k_event tracing in tests which seem completely unrelated to my PR.
I did suppress the errors by defining void macros, which is rather a workaround.

Will those tests not be executed frequently for those problems to surface?
Cannot see how it is related to my changes.

Any hint?

romkell committed:
kernel: make ENTRY(CONFIG_KERNEL_ENTRY) default

Make ENTRY(CONFIG_KERNEL_ENTRY) the default and
set CONFIG_KERNEL_ENTRY in KConfig.defconfig platform
or SOC specific (exluding expressifs with __appcpu_start
as entry symbol).

Signed-off-by: Roman Kellner <rkellner@baumer.com>
